lib/StackAnalysis/ABI.cpp

```diff
@@ -5,308 +5,666 @@
 
-  auto VoidKind = model::PrimitiveTypeKind::Void;
-  auto PrimitiveVoid = TheBinary.getPrimitiveType(VoidKind, 0);
-  QualifiedType Void{ PrimitiveVoid, {} };
+  model::RawFunctionType Result;
+  for (auto [ArgumentIndex, ArgumentStorage] : *Arguments) {
+    const auto &ArgumentType = Original.Arguments.at(ArgumentIndex).Type;
+    if (!ArgumentStorage.Registers.empty()) {
+      // Handle the registers
+      auto OriginalName = Original.Arguments.at(ArgumentIndex).CustomName;
+      for (size_t Index = 0; auto Register : ArgumentStorage.Registers) {
+        auto FinalName = OriginalName;
+        if (ArgumentStorage.Registers.size() > 1 && !FinalName.empty())
+          FinalName += "_part_" + std::to_string(++Index) + "_out_of_"
+                       + std::to_string(ArgumentStorage.Registers.size());
+        model::NamedTypedRegister Argument(Register);
+        Argument.Type = chooseArgumentType(ArgumentType,
+                                           Register,
+                                           ArgumentStorage.Registers,
+                                           TheBinary);
+        Argument.CustomName = FinalName;
+        Result.Arguments.insert(Argument);
+      }
+    }
 
-  model::CABIFunctionType Result;
-  Result.ABI = SystemV_x86_64;
-
-  //
-  // Build return type
-  //
-  QualifiedType ReturnType;
-
-  auto ReturnValuesCount = AnalysisResult.ReturnValues;
-  if (ReturnValuesCount == 0) {
-    ReturnType = Void;
-  } else if (ReturnValuesCount == 1) {
-    ReturnType = Generic64;
-  } else {
-    auto NewType = makeType<StructType>();
-    auto *MultipleReturnValues = llvm::cast<StructType>(NewType.get());
-    MultipleReturnValues->Size = ReturnValuesCount * 8;
-    for (uint64_t I = 0; I < AnalysisResult.ReturnValues; ++I) {
-      StructField NewField;
-      NewField.Offset = I * 8;
-      NewField.Type = Generic64;
-      MultipleReturnValues->Fields.insert(std::move(NewField));
+    if (ArgumentStorage.SizeOnStack != 0) {
+      // Handle the stack
+      auto ArgumentIterator = Original.Arguments.find(ArgumentIndex);
+      revng_assert(ArgumentIterator != Original.Arguments.end());
+      const model::Argument &Argument = *ArgumentIterator;
+
+      auto MaybeSize = Argument.Type.size();
+      revng_assert(MaybeSize.has_value());
+      size_t Size = *MaybeSize;
+
+      // TODO: handle stack arguments properly.
+      // \note: different ABIs could use different stack types.
+      // \sa: `clrcall` ABI.
+
+      constexpr size_t Alignment = CallingConvention::StackAlignment;
+      size_t Aligned = Size + Alignment - 1
+                       - (Size + Alignment - 1) % Alignment;
+
+      if constexpr (CallingConvention::UsePointerToCopyForStackArguments)
+        if (Aligned > CallingConvention::StackAlignment)
+          Result.FinalStackOffset += CallingConvention::StackAlignment;
+        else
+          Result.FinalStackOffset += Aligned;
+      else
+        Result.FinalStackOffset += Aligned;
```

Comment:
Uhm, what is your understanding of `FinalStackOffset`?
It should be:
* `0` for ABIs using the link register (i.e., return address not on the stack) and in which it's the caller's responsibility to cleanup the stack.
* GPR-size for ABIs that push the return address on the stack.
* Size of the stack arguments for ABIs where it's the callee's responsibility to cleanup the stack.

I'm not sure the stack alignment assumptions are relevant here. Maybe we can drop `CallingConvTrait::StackAlignment` altogether?

Reply:
For now I'm just using this extra variable as a temporary storage for "the combined size (including stack alignment) of all the arguments passed using the stack" which is definitely not correct. So no actual relation to any actual offsets.

As for dropping `CallingConvTrait::StackAlignment`, maybe that's indeed the right call, BUT this could maybe be helpful extra information for the situations when we try to parse/pass stack arguments.

Reply:
> For now I'm just using this extra variable as a temporary storage for "the combined size (including stack alignment) of all the arguments passed using the stack" which is definitely not correct. So no actual relation to any actual offsets.

Uhm, OK, let's not write that in the model then.

> As for dropping CallingConvTrait::StackAlignment

I'm hesitant to leave unused code, it tends to go bitrot.

Reply:
This is gone.